### PR TITLE
Add govulncheck Makefile target and per-CVE check script

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:

Adds tooling to detect CVE false positives from module-level scanners like Renovate.

Renovate flags CVEs at the Go **module** level, but the vulnerable **package/function** may not be compiled into kubemacpool. Go's official `govulncheck` [0] performs source-level call graph analysis and only reports CVEs where the vulnerable code is actually reachable.

This PR adds:
- `make govulncheck` — runs `govulncheck ./...` to scan all vulnerabilities at once
- `hack/check-cve.sh <CVE-ID>` — checks a specific CVE and reports whether it's a false positive

Example:
```
$ hack/check-cve.sh CVE-2025-22868
Scanning for CVE-2025-22868...
NOT AFFECTED: CVE-2025-22868 — false positive (vulnerable code is not reachable)
```

[0] https://go.dev/doc/security/vuln/

**Special notes for your reviewer**:

Tested across all stable branches (release-0.39 through release-0.45). All 12 open Renovate security PRs' CVEs were confirmed as false positives by govulncheck.

**Release note**:
```release-note
NONE
```

## Summary by CodeRabbit

* **New Features**
  * Added a local vulnerability scanning workflow with two commands: one to run scans and one to parse results.
  * Introduced a CVE-check script that accepts a `CVE-YYYY-NNNNN` input and reports whether it is **AFFECTED**, **NOT AFFECTED** (not reachable), or **NOT FOUND**.
* **Bug Fixes**
  * Improved result parsing to tolerate mixed scan output (non-JSON lines) and handle parsing edge cases more reliably.
* **Tests**
  * Added automated test coverage for CVE classification scenarios and parser robustness.

### CODE REVIEW

File: hack/check-cve.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+if echo "${OUTPUT}" | grep -q "${CVE_ID}"; then
+    echo "AFFECTED: ${CVE_ID} — vulnerable code is reachable"
+    echo ""
+    echo "${OUTPUT}" | grep -B2 -A5 "${CVE_ID}"
+    exit 1
+else
+    echo "NOT AFFECTED: ${CVE_ID} — false positive (vulnerable code is not reachable)"
+    exit 0
+fi
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply hack/check-cve.sh]: addressed comment. please PTAL again

**Comment by RamLavi** [comment]
[Change](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/dd3e1a17cf5b676da69aeb1242847d70bc39f981..d805dc8d54a2e165e0f733f13076c59a2735c2a8): use json format of govuln tool

File: hack/parse-govulncheck.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+            if "osv" in obj:
+                aliases = obj["osv"].get("aliases", [])
+                if cve in aliases:
+                    osv_ids_for_cve.add(obj["osv"]["id"])
+
+            if "finding" in obj:
+                finding_osv_ids.add(obj["finding"]["osv"])
+        except json.JSONDecodeError:
+            pos += match.start() + 1
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply hack/parse-govulncheck.py]: fixed. PTAL

**Comment by RamLavi** [comment]
[Change](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/d805dc8d54a2e165e0f733f13076c59a2735c2a8..4c91e7a29ce44db6f52eb02f9165f4575c2d3288): fixed a bug in counter

File: hack/check-cve.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        exit 1
+        ;;
+    NOT_REACHABLE)
+        echo "NOT AFFECTED: ${CVE_ID} — false positive (module is a dependency but vulnerable code is not reachable)"
+        exit 0
+        ;;
+    NOT_FOUND)
+        echo "NOT AFFECTED: ${CVE_ID} — false positive (vulnerable code is not reachable)"
+        exit 0
+        ;;
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply hack/check-cve.sh:46]: fixed

File: Makefile [hunk 4]
```diff
@@ -116,6 +116,21 @@ lint-fix: $(GOLANGCI_LINT_BIN)
...
+govulncheck: $(GOVULNCHECK_BIN)
+	@$(GOVULNCHECK_BIN) -format json ./...
+
+GOVULNCHECK_PARSER_BIN := $(BIN_DIR)/govulncheck-parser
+
+$(GOVULNCHECK_PARSER_BIN): $(GO)
+	$(GO) build -o $(GOVULNCHECK_PARSER_BIN) ./tools/govulncheck-parser
+
+govulncheck-parser: $(GOVULNCHECK_PARSER_BIN)
+
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply Makefile:135]: Fixed. Added source file dependency via `$(wildcard tools/govulncheck-parser/*.go)`.

File: tools/govulncheck-parser/main_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+			result, err := parser.CheckCVE(strings.NewReader(tt.input), tt.cveID)
+			if err != nil {
+				t.Fatalf("CheckCVE returned unexpected error: %v", err)
+			}
+			if result != tt.expected {
+				t.Errorf("CheckCVE(%q) = %q, want %q", tt.cveID, result, tt.expected)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply tools/govulncheck-parser/main_test.go]: Fixed. Refactored to Ginkgo v2 + Gomega with `DescribeTable`/`Entry`.

File: tools/govulncheck-parser/main.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+				if alias == cveID {
+					osvIDsForCVE[rec.OSV.ID] = true
+				}
+			}
+		}
+
+		if rec.Finding != nil && rec.Finding.OSV != "" {
+			findingOSVIDs[rec.Finding.OSV] = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply tools/govulncheck-parser/main.go:77]: Non-JSON lines are already stripped by `stripNonJSON()` before the decoder sees anything. If `json.Decoder` still errors on the cleaned input, it means govulncheck produced genuinely malformed JSON — that's a broken tool, not something we should silently recover from. Breaking and returning what we have is safer than guessing where to resume in corrupted output.

File: tools/govulncheck-parser/main.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+	return ResultNotFound, nil
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <CVE-ID>\n", os.Args[0])
+		os.Exit(2)
+	}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply tools/govulncheck-parser/main.go:101]: Fixed. Extracted `const expectedArgs = 2`.

File: tools/govulncheck-parser/main.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	return ResultNotFound, nil
+}
+
+const expectedArgs = 2
+
+func main() {
+	if len(os.Args) != expectedArgs {
+		fmt.Fprintf(os.Stderr, "Usage: %s <CVE-ID>\n", os.Args[0])
+		os.Exit(2)
+	}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply tools/govulncheck-parser/main.go]: Pedantic, but fixed.

File: tools/govulncheck-parser/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply tools/govulncheck-parser/main.go:77]: Acknowledged — it's a conscious tradeoff. In practice, `govulncheck -format json` produces well-formed NDJSON, and `stripNonJSON()` handles the only real-world noise (make output lines). A mid-stream decode error would mean govulncheck itself is broken, at which point the scan results are unreliable regardless of how we handle the error. We'd rather return a partial result than fabricate a failure for a scenario that doesn't happen in practice.

**Comment by RamLavi** [comment]
Thanks @qinqon 

Once I have ACK from Mirco, I'll run this script on all current open CVE PRs on the repo, and add this tool for CNAO as well
